src/eduid_userdb/actions/tou/user.py

```diff
@@ -44,33 +46,42 @@ class ToUUser(User):
     the eduid-actions plugin for ToU specific data.
 
     :param userid: user id
-    :type userid: bson.ObjectId
     :param eppn: eppn
-    :type eppn: str
     :param tou: ToU  list
-    :type tou: list
     :param data: userid, eppn and tou
-    :type data: dict
     :param raise_on_unknown: whether to raise an exception if
                              there is unknown data in the data dict
-    :type raise_on_unknown: bool
     """
 
-    def __init__(self, userid=None, eppn=None, tou=None, data=None, raise_on_unknown=True):
+    def __init__(
+        self,
+        userid: Optional[Union[str, bson.ObjectId]] = None,
+        eppn: Optional[str] = None,
+        tou: Optional[list] = None,
+        data: Optional[dict] = None,
+        raise_on_unknown: bool = True,
+        called_directly: bool = True,
+    ):
         """
         """
         if data is None:
             data = {'_id': userid, 'eduPersonPrincipalName': eppn, 'tou': tou}
 
+        User.__init__(self, data, raise_on_unknown=raise_on_unknown, called_directly=called_directly)
+
+    def check_or_use_data(self):
+        """
+        Check that the provided data dict contains all needed keys.
+        """
+        data = self._data_in
         if '_id' not in data or data['_id'] is None:
             raise UserMissingData('Attempting to record a ToU acceptance ' 'for an unidentified user.')
         if 'eduPersonPrincipalName' not in data or data['eduPersonPrincipalName'] is None:
-            raise UserMissingData('Attempting to record a ToU acceptance ' 'for a user without eppn.')
+            if 'eppn' not in data or data['eppn'] is None:
```

Comment:
It seems like you add tolerance for 'eppn' in data here, where it wasn't supported earlier. Why?

I also fail to see where data['eppn'] would actually be used, since User.from_dict doesn't "normalise" it, and User.__init__() only looks for data['eduPersonPrincipalName'].

Reply:
True. The why - it's a relic of a point at which I was putting the interface between the 2 naming conventions on the other side of this code - and a bug. Fixed.